geomstats/learning/riemannian_mean_shift.py

```diff
@@ -0,0 +1,147 @@
+"""Riemannian mean-shift clustering
+
+Lead author: Nina Miolane"""
+
+import joblib
+import numpy as np
+from sklearn.base import BaseEstimator, ClusterMixin
+
+from geomstats.learning.frechet_mean import FrechetMean
+
+
+class RiemannianMeanShift(ClusterMixin, BaseEstimator):
+    def __init__(self, manifold, metric, bandwidth, tol=1e-2, **FrechetMean_kwargs):
+        """Class for Riemannian Mean Shift algorithm on manifolds.
+        Mean Shift is a procedure for locating the maxima - the modes of a
+        density function given discrete data sampled from that function. It is
+        an iterative method for finding the centers of a collection of clusters.
+
+        Following implementation assumes a numpy backend and a flat kernel method.
+
+        Parameters
+        ----------
+        manifold : object of class RiemannianManifold
+            The Geomstats Riemannian manifold on which the
+            Riemannian Mean Shift algorithm is to applied.
+        metric : object of class RiemannianMetric
+            The Geomstats Riemannian metric associated to the space used.
+        bandwidth : size of neighbourhood around each center
+            All points in 'bandwidth' size around center are considered for
+            calculating new mean centers.
+        tot : float, stopping condition
+            Computation of subsequent mean centers is stopped when the distance
+            between them is less than 'tot'
+        mean : mean method used to calculate the new center of a cluster
+        """
+        self.manifold = manifold
+        self.metric = metric
+        self.bandwidth = bandwidth
+        self.tol = tol
+        self.mean = FrechetMean(self.metric, **FrechetMean_kwargs)
+        self.centers = None
+
+    def __intersets_distances(self, points_A, points_B, n_jobs=1, **joblib_kwargs):
+        """
+        Parallel computation of distances between two sets of points.
+        """
+        n_A, n_B = points_A.shape[0], points_B.shape[0]
+
+        @joblib.delayed
+        @joblib.wrap_non_picklable_objects
+        def pickable_dist(x, y):
+            return self.metric.dist(x, y)
+
+        pool = joblib.Parallel(n_jobs=n_jobs, **joblib_kwargs)
+        out = pool(
+            pickable_dist(points_A[i, :], points_B[j, :])
+            for i in range(n_A)
+            for j in range(n_B)
+        )
+
+        # assuming numpy backend
+        return np.array(out).reshape((n_A, n_B))
+
+    def fit(
+        self,
+        points,
+        n_centers,
+        n_jobs=1,
+        max_iter=100,
+        init_centers="from_points",
+        kernel="flat",
```

Comment:
Following scikit learn's API, these parameters should be attributes of the object and passed to init.
`points` should be named `X`, and a dummy `y=None` argument needs to be added for consistency with other estimators.

Reply:
Thank you Nicolas for your comments. I have raised a new PR with the changes that you have suggested. Please review them at your convenience and let me know of more changes or improvements.